kernel: add DesignWare SPI MMIO controller module

AlbrechtL opened the pull request:
This patch enables the SPI controller on the Raspberry Pi 5 to support external SPI devices. Functionality has been tested with a Microchip ENC28J60 Ethernet controller.

This PR is an outcome of https://github.com/openwrt/openwrt/issues/18034.

hauke commented on target/linux/bcm27xx/bcm2712/config-6.12:
```diff
@@ -583,6 +583,14 @@ CONFIG_SPARSEMEM_EXTREME=y
 CONFIG_SPARSEMEM_VMEMMAP=y
 CONFIG_SPARSEMEM_VMEMMAP_ENABLE=y
 CONFIG_SPARSE_IRQ=y
+CONFIG_SPI=y
+# CONFIG_SPI_BCM2835AUX is not set
+CONFIG_SPI_DESIGNWARE=y
+# CONFIG_SPI_DW_DMA is not set
+CONFIG_SPI_DW_MMIO=y
+# CONFIG_SPI_DW_PCI is not set
+CONFIG_SPI_DYNAMIC=y
+CONFIG_SPI_MASTER=y
```
I would prefer if you build this as a loadable kernel module in `package/kernel/linux/modules/spi.mk`.

AlbrechtL replied:
Thanks for reviewing! I will do it.

Noltari reviewed (changes requested):
As @hauke suggested, this PR should add new generic kernel modules for DesignWare SPI and therefore it shouldn't be focused on the RPi 5 since this a generic issue that should affect other targets as well.

AlbrechtL commented:
> As @hauke suggested, this PR should add new generic kernel modules for DesignWare SPI and therefore it shouldn't be focused on the RPi 5 since this a generic issue that should affect other targets as well.

Thanks for pointing it out! I already started to put it in `target/linux/bcm27xx/modules/spi.mk`. Now, I will use `package/kernel/linux/modules/spi.mk`.

AlbrechtL commented:
I just updated the code, please review it.

But I have the issue that the kernel module `spi-dw-mmio` is not loaded automatically. When executing `modprobe spi-dw-mmio` everything is working fine.

@Noltari, @hauke  Do you know why?

hauke commented on package/kernel/linux/modules/spi.mk:
```diff
@@ -73,3 +73,27 @@ define KernelPackage/spi-dev/description
+  FILES:=\
+    $(LINUX_DIR)/drivers/spi/spi-dw.ko \
+    $(LINUX_DIR)/drivers/spi/spi-dw-mmio.ko
+  AUTOLOAD:=$(call AutoLoad,spi-dw-mmio)
```
```suggestion
  AUTOLOAD:=$(call AutoProbe,spi-dw-mmio)
```
This should fix loading the module.

`AutoLoad` needs a order as second parameter, `AutoProbe` will automatically load all dependencies and is preferred.

AlbrechtL replied:
Thanks for the explanation! I just changed it.

Noltari commented on package/kernel/linux/modules/spi.mk:
```diff
@@ -73,3 +73,27 @@ define KernelPackage/spi-dev/description
 $(eval $(call KernelPackage,spi-dev))
+
+
+define KernelPackage/spi-dw-mmio
+  SUBMENU:=$(SPI_MENU)
+  TITLE:=DesignWare SPI controller driver (MMIO)
+  KCONFIG:=\
+    CONFIG_SPI=y \
+    CONFIG_SPI_DESIGNWARE \
+    CONFIG_SPI_DW_DMA=n \
+    CONFIG_SPI_DW_MMIO=y \
+    CONFIG_SPI_DW_PCI=n \
+    CONFIG_SPI_DYNAMIC=y \
+    CONFIG_SPI_MASTER=y
+  FILES:=\
+    $(LINUX_DIR)/drivers/spi/spi-dw.ko \
+    $(LINUX_DIR)/drivers/spi/spi-dw-mmio.ko
+  AUTOLOAD:=$(call AutoProbe,spi-dw-mmio)
+endef
```
I think that we should add another kernel package for spi-dw and let spi-dw-mmio depend on it.
Something like:
```
define KernelPackage/spi-dw
  SUBMENU:=$(SPI_MENU)
  TITLE:=DesignWare SPI controller driver (core)
  KCONFIG:=\
    CONFIG_SPI=y \
    CONFIG_SPI_DESIGNWARE \
    CONFIG_SPI_DYNAMIC=y \
    CONFIG_SPI_MASTER=y
  FILES:=\
    $(LINUX_DIR)/drivers/spi/spi-dw.ko
  AUTOLOAD:=$(call AutoProbe,spi-dw)
endef
```

```
define KernelPackage/spi-dw-mmio
  SUBMENU:=$(SPI_MENU)
  TITLE:=DesignWare SPI controller driver (MMIO)
  DEPENDS:=+kmod-spi-dw-mmio
  KCONFIG:=\
    CONFIG_SPI_DW_MMIO
  FILES:=\
    $(LINUX_DIR)/drivers/spi/spi-dw-mmio.ko
  AUTOLOAD:=$(call AutoProbe,spi-dw-mmio)
endef
```

And then also add the following to target/linux/generic/config-6.12:

```
# CONFIG_SPI_DW_BT1 is not set
# CONFIG_SPI_DW_BT1_DIRMAP is not set
# CONFIG_SPI_DW_DMA is not set
# CONFIG_SPI_DW_MMIO is not set
# CONFIG_SPI_DW_PCI is not set
```

AlbrechtL replied:
Implemented in https://github.com/openwrt/openwrt/commit/9ab8bd79e36378a3230df10d502451c035130877.

hauke commented:
There are formal problems in this PR:
```
Commit subject line MUST start with '<area>: ' (Add support for DesignWare SPI controller driver with MMIO interface.)
```

AlbrechtL committed:
kernel: add support for DesignWare SPI MMIO controller

Add kernel packages for DesignWare SPI core and MMIO controllers.
This is needed for the RP1 SoC found on RPi 5 devices.

Tested with a Microchip ENC28J60 Ethernet controller on a RPi 5.

Signed-off-by: Albrecht Lohofener <albrechtloh@gmx.de>
Link: https://github.com/openwrt/openwrt/pull/19049
Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

AlbrechtL commented:
> There are formal problems in this PR

Fixed in https://github.com/openwrt/openwrt/commit/9ab8bd79e36378a3230df10d502451c035130877.

AlbrechtL commented:
Since this PR in generic now. Does it make sense to issue another PR to enable the DesignWare SPI controller by default for the Raspberry Pi 5 target?

Noltari commented:
> Since this PR in generic now. Does it make sense to issue another PR to enable the DesignWare SPI controller by default for the Raspberry Pi 5 target?

We're not enabling SPI by default on any other bcm27xx subtarget so I wouldn't do that for bcm2712 either.
In the end the user can manually install/select the SPI modules by himself.
What we could do is add a conditional dependency on spi-bcm2835 to spi-dw-mmio if target is bcm2712.

AlbrechtL commented:
>  What we could do is add a conditional dependency on spi-bcm2835 to spi-dw-mmio if target is bcm2712.

That would also an idea. Because I'm a little bit uncertain that users know that for bcm2712 SPI support the DesignWare controller driver is necessary.

>  What we could do is add a conditional dependency on spi-bcm2835 to spi-dw-mmio if target is bcm2712.

This would be a new PR after this one is done, right?

Noltari commented:
> >  What we could do is add a conditional dependency on spi-bcm2835 to spi-dw-mmio if target is bcm2712.
> 
> 
> 
> That would also an idea. Because I'm a little bit uncertain that users know that for bcm2712 SPI support the DesignWare controller driver is necessary. But it is interesting that the DesignWare controller is working even when `spi-bcm2835` is not loaded. It seems that also `spidev` is not needed for `spi_dw`.

Ah, excuse me but I thoght you needed both.
In that case no, we shouldn't add the dependency.

> See:
> 
> ```
> 
> root@OpenWrt:~# lsmod | grep spi
> 
> spi_dw                 16384  1 spi_dw_mmio
> 
> spi_dw_mmio            12288  0 
> 
> spidev                 20480  0 
> 
> ```
> 
> 
> 
> 
> 
> >  What we could do is add a conditional dependency on spi-bcm2835 to spi-dw-mmio if target is bcm2712.
> 
> 
> 
> This would be a new PR after this one is done, right?

Yes.

AlbrechtL commented:
>  In that case no, we shouldn't add the dependency.

So, that means that users have to know that `spi_dw_mmio` is the correct one for external SPI at the Raspberry Pi 5. Maybe it is worth to document this at https://openwrt.org/toh/raspberry_pi_foundation/raspberry_pi#using_i2c_and_spi.

AlbrechtL commented:
> We're not enabling SPI by default on any other bcm27xx subtarget so I wouldn't do that for bcm2712 either.
In the end the user can manually install/select the SPI modules by himself.

Is there a specific reason for this decision? I understand that OpenWrt is trying to keep the image as small as possible to fit into tiny flash storages. But we have this situation not for the Raspberry Pis. The smallest SD-cards you can buy are 1-4 Gigabyte. In my opinion, all Raspberry Pi HW drivers should be installed by default (incl. SPI, I2C etc.). I don't think that this would blow up the image size.

AlbrechtL commented:
@Noltari Thanks for your support!

brocaar commented:
@Noltari will this be cherry-picked into the `openwrt-24.10` branch / what would be the process to integrate this into the `openwrt-24.10`?

Noltari commented:
> @Noltari will this be cherry-picked into the `openwrt-24.10` branch / what would be the process to integrate this into the `openwrt-24.10`?

Yes, I intend to cherry pick all the I2C/SPI changes related to RPi devices to OpenWrt 24.10 once they are all merged.
https://github.com/openwrt/openwrt/commit/82d20a0fbde937365f2ad6b4f9b754971c5c6ab2
https://github.com/openwrt/openwrt/commit/4987239913794811b8ac71ff141129e276565d6b
https://github.com/openwrt/openwrt/commit/84922a1e34f8fafeeebc3e6a25b93b95728c1706
https://github.com/openwrt/openwrt/pull/19086
https://github.com/Noltari/openwrt/tree/bcm27xx-i2c-spi-packages
